### PR TITLE
Gradle hide empty generated folders

### PR DESCRIPTION
**Pull request by lkishalmi**
When added the support of Generated source directories in Gradle project, the project view become a bit crowded, as the generated folders are always generated by Gradle.

Here there are two empty folder one for java and one for groovy.
image

Same project with the new code.
image

While implementing the change, I discovered a bug which prevented the freshly opened project to listen on resource changes. That one is the fist commit.

### CODE REVIEW

File: extide/gradle/src/org/netbeans/modules/gradle/api/NbGradleProject.java [hunk 1]
```diff
@@ -255,34 +255,37 @@ private void fireProjectReload() {
...
-            resources.addAll(pvd.getWatchedResources());
-        }
-        for (File resource : resources) {
-            try {
-                FileUtil.addFileChangeListener(FCHSL, resource);
-            } catch (IllegalArgumentException ex) {
-                assert false : "Something is wrong with the resource handling";
+        if ((project.getAimedQuality() == Quality.FALLBACK) && !elevateQuality) return;
+        synchronized (resources) {
+            Collection<? extends WatchedResourceProvider> all
```
Comment (sdedic) [inline extide/gradle/src/org/netbeans/modules/gradle/api/NbGradleProject.java:278]: I'd clear `resources` at the start.

File: java/gradle.java/src/org/netbeans/modules/gradle/java/classpath/GradleSourcesImpl.java [hunk 2]
```diff
@@ -159,7 +160,7 @@ public synchronized SourceGroup[] getSourceGroups(String type) {
         for (SourceType st : stype) {
             Set<File> processed = new HashSet<>();
             for (String group : gradleSources.keySet()) {
-                Set<File> dirs = gradleSources.get(group).getSourceDirs(st);
+                Set<File> dirs = filterEmptyGeneratedDirs(gradleSources.get(group).getSourceDirs(st));
```
Comment (sdedic) [inline java/gradle.java/src/org/netbeans/modules/gradle/java/classpath/GradleSourcesImpl.java:163]: Is it OK to filter the empty dirs from the data layer ? I'd rather leave them discoverable (as they are valid source roots), but filter them from the UI only.

Comment (lkishalmi) [reply java/gradle.java/src/org/netbeans/modules/gradle/java/classpath/GradleSourcesImpl.java]: Good insight! Thanks! Moving it to the UI!

File: java/gradle.java/src/org/netbeans/modules/gradle/java/nodes/SourcesNodeFactory.java [hunk 3]
```diff
@@ -79,7 +97,8 @@ public List<SourceGroup> keys() {
             ret.addAll(Arrays.asList(srcs.getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA)));
             ret.addAll(Arrays.asList(srcs.getSourceGroups(GradleSourcesImpl.SOURCE_TYPE_KOTLIN)));
             ret.addAll(Arrays.asList(srcs.getSourceGroups(JavaProjectConstants.SOURCES_TYPE_RESOURCES)));
-            ret.addAll(Arrays.asList(srcs.getSourceGroups(GradleSourcesImpl.SOURCE_TYPE_GENERATED)));
+            generatedGroups = Arrays.asList(srcs.getSourceGroups(GradleSourcesImpl.SOURCE_TYPE_GENERATED));
```
Comment (sdedic) [inline java/gradle.java/src/org/netbeans/modules/gradle/java/nodes/SourcesNodeFactory.java:100]: nitpick: `generatedGroups` access is unsynchronized; here, the value should be assigned first to local, then to generatedGroups + added to ret, to be consistent (within this thread). `generatedGroups` could be possibly `volatile`

**Review by sdedic — APPROVED** [review]

**Pull request merged**